mystery-deckbuilder/Assets/Scripts/DialogueOverlay/DialogueBoxManager.cs

```diff
@@ -0,0 +1,187 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+/*
+ * The singleton class responsible for the following:
+ * Instantiating DialogueBox GameObjects
+ * -Starting dialogues
+ * -traversing a given dialogue tree
+ * -Commanding the instantiated DialogueBox to display the next sentence and show the option box
+ * -Ending the dialogue and commanding the DialogueBox to destroy itself
+*/
+public class DialogueBoxManager : MonoBehaviour
+{
+
+    public static DialogueBoxManager Instance {get; private set; } //a static instance of itself
+
+    private DialogueTree _dialogueTree;
+    private IDialogueNode _currentNode;
+
+    private Queue<string> _sentences = new();
+
+    [SerializeField] private GameObject _dialogueBoxPrefab;
+    private GameObject _dialogueBox;
+
+    /* Since this is a singleton class, we destroy all other instances of it that aren't this one */
+    private void Awake()
+    {
+        if (Instance != null && Instance != this)
+        {
+            Destroy(this);
+        }
+        else
+        {
+            Instance = this;
+        }
+    }
+
+    // Start is called before the first frame update
+    void Start()
+    {
+        _sentences = new Queue<string>();
+    }
+
+    /* A built-in dialogue to be called by the button in the DialogueTest scene 
+     * for the purpose of demonstration. Uses Gabe's example dialogue tree.
+    */
+    public void TestDialogueStart() 
+    {
+        StartDialogue(transform.GetComponent<ExampleDialogue>().dialogueTree);
+    }
+
+    /* Initiates a new dialogue by instantiating a DialogueBox, adding its sentences to the queue
+     * setting the name of the DialogueBox, and displaying the next sentence (in this case the first one)
+     */
+    public void StartDialogue(DialogueTree newDialogueTree)
+    {
+        //the current dialogue must be ended before starting a new one
+        if (_dialogueBox != null)
+        {
+            Debug.LogError("Error: tried to start a new dialogue when one was already in session");
+            return;
+        }
+
+        _dialogueTree = newDialogueTree;
+
+        //instantiate the dialogue box prefab
+        _dialogueBox = Instantiate(_dialogueBoxPrefab, new Vector3(0, -7, 0), Quaternion.identity);
+
+        GoToNode(_dialogueTree.root); //we start at the root node of the dialogue tree
+    }
+
+    /* Moves to the next node in the dialogue tree
+    */
+    public void NextNode()
+    {
+        _currentNode = _currentNode.Next();
+
+        //end dialogue if we've reached the end
+        if (_currentNode == null)
+        {
+            EndDialogue();
+            return;
+        }
+
+        StopAllCoroutines(); //stop displaying text, in case player clicks next while text still writing
+
+        if (_currentNode.NodeType() != "option") //if a normal dialogue node
+        {
+
+            _dialogueBox.GetComponent<DialogueBox>().SetName(_currentNode.NodeType()); //change this later to grab name of NPC
+            
```

Comment:
NIT: for sake of being ctrl+f - able, it can be nice to always write TODO on things that should be changed later